math/mathcore/inc/Fit/FitUtil.h

```diff
@@ -259,1212 +207,174 @@ namespace FitUtil {
-            if (!normalizeFunc) {
-               IntegralEvaluator<IModelFunctionTempl<T>> igEval(func, p, true);
-               std::vector<double> xmin(data.NDim());
-               std::vector<double> xmax(data.NDim());
-
-               // compute integral in the ranges where is defined
-               if (data.Range().Size() > 0) {
-                  nuTot = 0;
-                  for (unsigned int ir = 0; ir < data.Range().Size(); ++ir) {
-                     data.Range().GetRange(&xmin[0], &xmax[0], ir);
-                     nuTot += igEval.Integral(xmin.data(), xmax.data());
-                  }
-               } else {
-                  // use (-inf +inf)
-                  data.Range().GetRange(&xmin[0], &xmax[0]);
-                  // check if funcition is zero at +- inf
-                  T xmin_v, xmax_v;
-                  vecCore::Load<T>(xmin_v, xmin.data());
-                  vecCore::Load<T>(xmax_v, xmax.data());
-                  if (vecCore::ReduceAdd(func(&xmin_v, p)) != 0 || vecCore::ReduceAdd(func(&xmax_v, p)) != 0) {
-                     MATH_ERROR_MSG("FitUtil::EvaluateLogLikelihood", "A range has not been set and the function is not zero at +/- inf");
-                     return 0;
-                  }
-                  nuTot = igEval.Integral(&xmin[0], &xmax[0]);
-               }
-
-               // force to be last parameter value
-               //nutot = p[func.NDim()-1];
-               if (iWeight != 2)
-                  extendedTerm = - nuTot;  // no need to add in this case n log(nu) since is already computed before
-               else {
-                  // case use weight square in likelihood : compute total effective weight = sw2/sw
-                  // ignore for the moment case when sumW is zero
-                  extendedTerm = - (sumW2 / sumW) * nuTot;
-               }
-
-            } else {
-               nuTot = norm;
-               extendedTerm = - nuTot + double(n) *  ROOT::Math::Util::EvalLog(nuTot);
-               // in case of weights need to use here sum of weights (to be done)
-            }
-
-            logl += extendedTerm;
-         }
-
-#ifdef DEBUG_FITUTIL
-         std::cout << "Evaluated log L for parameters (";
-         for (unsigned int ip = 0; ip < func.NPar(); ++ip)
-            std::cout << " " << p[ip];
-         std::cout << ")  nll = " << -logl << std::endl;
-#endif
-         
-         return -logl;
-
-      }
+   // derivative with respect of the parameter to be integrated
+   template <class GradFunc = IGradModelFunction>
+   struct ParamDerivFunc {
+      ParamDerivFunc(const GradFunc &f) : fFunc(f), fIpar(0) {}
+      void SetDerivComponent(unsigned int ipar) { fIpar = ipar; }
+      double operator()(const double *x, const double *p) const { return fFunc.ParameterDerivative(x, p, fIpar); }
+      unsigned int NDim() const { return fFunc.NDim(); }
+      const GradFunc &fFunc;
+      unsigned int fIpar;
```

Comment:
Why have this as member instead of passing to `operator()`? Looks like it makes it harder to parallelize, since you have to `SetDerivComponent(n);` then calculate derivative, which then cannot happen concurrently using the same `ParamDerivFunc` object.